### PR TITLE
Adds new X-Effective-URL header to the xdebug plugin

### PR DESCRIPTION
**Pull request by jhiapple**
By adding "X-Effective-URL" to the request header "X-Debug",
the header "X-Effective-URL" header will be inserted to the
request headers with the effective URL value.

Exposes the TSHttpTxnEffectiveUrlStringGet() result, which is
the URL after the remap.

Note: the inserted header is intentionally X-prefixed for consistency
with the other headers inserted by the xdebug plugin.

### CODE REVIEW

**Comment by elsloo** [comment]
[approve ci]

**Comment by jhiapple** [comment]
[approve ci]

**Comment by zwoop** [comment]
[add to whitelist]

**Comment by zwoop** [comment]
[approve ci]

**Comment by jhiapple** [comment]
[approve ci]

**Comment by bryancall** [comment]
I will review it

**Comment by ywkaras** [comment]
Do we want this in 9.2 or not until 10?

**Comment by zwoop** [comment]
We can add this to 9.2.x.

File: plugins/xdebug/xdebug.cc [hunk 1]
```diff
@@ -376,6 +377,34 @@ InjectRemapHeader(TSHttpTxn txn, TSMBuffer buffer, TSMLoc hdr)
...
+
+  if (strval.ptr != nullptr && strval.len > 0) {
+    dst = FindOrMakeHdrField(buffer, hdr, "X-Effective-URL", lengthof("X-Effective-URL"));
+    if (dst != TS_NULL_MLOC) {
+      TSReleaseAssert(TSMimeHdrFieldValueStringInsert(buffer, hdr, dst, -1 /* idx */, strval.ptr, strval.len) == TS_SUCCESS);
+    }
+  }
+
+  if (dst != TS_NULL_MLOC) {
+    TSHandleMLocRelease(buffer, hdr, dst);
```
Comment (ywkaras) [inline plugins/xdebug/xdebug.cc:402]: You could move this between lines 397 and 398, and eliminate this if statement.  Once you did that, you could declare dst on line 395 and delete line 303.

**Comment by jhiapple** [comment]
> On 14. Jul 2021, at 0.40, Walt Karas ***@***.***> wrote:
> 
> 
> @ywkaras commented on this pull request.
> 
> In plugins/xdebug/xdebug.cc <https://github.com/apache/trafficserver/pull/7931#discussion_r669129269>:
> 
> > +    int len;
> +  } strval = {nullptr, 0};
> +
> +  TSDebug("xdebug", "attempting to inject X-Effective-URL header");
> +
> +  strval.ptr = TSHttpTxnEffectiveUrlStringGet(txn, &strval.len);
> +
> +  if (strval.ptr != nullptr && strval.len > 0) {
> +    dst = FindOrMakeHdrField(buffer, hdr, "X-Effective-URL", lengthof("X-Effective-URL"));
> +    if (dst != TS_NULL_MLOC) {
> +      TSReleaseAssert(TSMimeHdrFieldValueStringInsert(buffer, hdr, dst, -1 /* idx */, strval.ptr, strval.len) == TS_SUCCESS);
> +    }
> +  }
> +
> +  if (dst != TS_NULL_MLOC) {
> +    TSHandleMLocRelease(buffer, hdr, dst);
> You could move this between lines 397 and 398, and eliminate this if statement. Once you did that, you could declare dst on line 395 and delete line 303.
> 
> 

Tighter scope nice, so done in https://github.com/apache/trafficserver/pull/7931/commits/60252d542c5b5d49b81d135ead0c631bab03f9ed

**Review by ywkaras — APPROVED** [review]

**Pull request merged**